src/uct/ugni/rdma/ugni_rdma_iface.c

```diff
@@ -0,0 +1,272 @@
+                          "resource desc");
+    if (NULL == resources) {
+      ucs_error("Failed to allocate memory");
+      num_devices = 0;
+      resources = NULL;
+      rc = UCS_ERR_NO_MEMORY;
+      goto error;
+    }
+
+    for (i = 0; i < job_info.num_devices; i++) {
+        uct_ugni_device_get_resource(&devs[i], &resources[i]);
+    }
+
+error:
+    *num_resources_p = num_devices;
+    *resource_p      = resources;
+    pthread_mutex_unlock(&uct_ugni_global_lock);
+
+    return rc;
+}
+
+static ucs_status_t uct_ugni_rdma_iface_query(uct_iface_h tl_iface, uct_iface_attr_t *iface_attr)
+{
+    uct_ugni_rdma_iface_t *iface = ucs_derived_of(tl_iface, uct_ugni_rdma_iface_t);
+
+    memset(iface_attr, 0, sizeof(uct_iface_attr_t));
+    iface_attr->cap.put.max_short      = iface->config.fma_seg_size;
+    iface_attr->cap.put.max_bcopy      = iface->config.fma_seg_size;
+    iface_attr->cap.put.max_zcopy      = iface->config.rdma_max_size;
+    iface_attr->cap.get.max_bcopy      = iface->config.fma_seg_size - 8; /* alignment offset 4 (addr)+ 4 (len)*/
+    iface_attr->cap.get.max_zcopy      = iface->config.rdma_max_size;
+    iface_attr->iface_addr_len         = sizeof(uct_sockaddr_ugni_t);
+    iface_attr->ep_addr_len            = 0;
+    iface_attr->cap.flags              = UCT_IFACE_FLAG_PUT_SHORT |
+                                         UCT_IFACE_FLAG_PUT_BCOPY |
+                                         UCT_IFACE_FLAG_PUT_ZCOPY |
+                                         UCT_IFACE_FLAG_ATOMIC_CSWAP64 |
+                                         UCT_IFACE_FLAG_ATOMIC_FADD64  |
+                                         UCT_IFACE_FLAG_ATOMIC_ADD64   |
+                                         UCT_IFACE_FLAG_GET_BCOPY      |
+                                         UCT_IFACE_FLAG_GET_ZCOPY      |
+                                         UCT_IFACE_FLAG_CONNECT_TO_IFACE;
+    return UCS_OK;
+}
+
+
+static UCS_CLASS_CLEANUP_FUNC(uct_ugni_rdma_iface_t)
+{
+    gni_return_t ugni_rc;
+
+    ucs_notifier_chain_remove(&self->super.super.worker->progress_chain,
+                              uct_ugni_progress, self);
+
+    if (!self->super.activated) {
+        /* We done with release */
+        return;
+    }
+
+    ucs_mpool_destroy(self->free_desc_get_buffer);
+    ucs_mpool_destroy(self->free_desc_get);
+    ucs_mpool_destroy(self->free_desc_famo);
+    ucs_mpool_destroy(self->free_desc_buffer);
+    ucs_mpool_destroy(self->free_desc);
+
+    /* TBD: Clean endpoints first (unbind and destroy) ?*/
+    ugni_rc = GNI_CqDestroy(self->super.local_cq);
+    if (GNI_RC_SUCCESS != ugni_rc) {
+        ucs_warn("GNI_CqDestroy failed, Error status: %s %d",
+                 gni_err_str[ugni_rc], ugni_rc);
+    }
+    ugni_rc = GNI_CdmDestroy(self->super.cdm_handle);
+    if (GNI_RC_SUCCESS != ugni_rc) {
+        ucs_warn("GNI_CdmDestroy error status: %s (%d)",
+                 gni_err_str[ugni_rc], ugni_rc);
+    }
```

Comment:
all the super calls - since it is allocated in base shell we split the cleanup to rdma cleanup and base cleanup, where base is responsible to release CQ and CDM handle